### PR TITLE
FINERACT-1958-Configuration-down-payment

### PR DESCRIPTION
**Pull request by ruchiD**
## Description
Added fields for down-payment configuration for loan product.
Validations
Swagger changes
Integration test

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java [hunk 1]
```diff
@@ -210,6 +210,12 @@ public class LoanProduct extends AbstractPersistableCustom {
     @Column(name = "overdue_days_for_repayment_event")
     private Integer overDueDaysForRepaymentEvent;
 
+    @Column(name = "enable_down_payment", nullable = false)
+    private boolean enableDownPayment;
+
+    @Column(name = "disbursed_amount_percentage_down_payment", scale = 2, precision = 5)
+    private BigDecimal disbursedAmountPercentageDownPayment;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java]: i am not too fond of this naming. Maybe disbursedAmountPercentage**For**DownPayment?

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java [hunk 2]
```diff
@@ -385,6 +391,11 @@ public static LoanProduct assembleFromJson(final Fund fund, final String loanTra
         final Integer overDueDaysForRepaymentEvent = command
                 .integerValueOfParameterNamed(LoanProductConstants.OVER_DUE_DAYS_FOR_REPAYMENT_EVENT);
 
+        final boolean enableDownPayment = command.parameterExists(LoanProductConstants.ENABLE_DOWN_PAYMENT)
+                && command.booleanPrimitiveValueOfParameterNamed(LoanProductConstants.ENABLE_DOWN_PAYMENT);
+        final BigDecimal disbursedAmountPercentageDownPayment = command
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java:389]: command.booleanPrimitiveValueOfParameterNamed is doing null check.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    public void setup() {
+        Utils.initializeRESTAssured();
+        this.requestSpec = new RequestSpecBuilder().setContentType(ContentType.JSON).build();
+        this.requestSpec.header("Authorization", "Basic " + Utils.loginIntoServerAndGetBase64EncodedAuthenticationKey());
+        this.responseSpec = new ResponseSpecBuilder().expectStatusCode(200).build();
+        this.loanTransactionHelper = new LoanTransactionHelper(this.requestSpec, this.responseSpec);
+    }
+
+    @Test
+    public void loanProductCreationWithDownPaymentConfigurationTest() {
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:68]: Can you please have the following extra tests?
- Down payment is not enabled, but percentage is set.
- Down payment is enabled, but percentage is 0 or 101
- Down payment was set and percentage also. After update loan product and turn off down payment. What is the result? What is the value of the percentage of down payment?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my questions!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java [hunk 4]
```diff
@@ -744,9 +745,38 @@ public void validateForCreate(final String json) {
...
     }
 
+    private void validateDownPaymentPercentage(Boolean enableDownPayment, DataValidatorBuilder baseDataValidator, JsonElement element) {
+        if (enableDownPayment) {
+            if (this.fromApiJsonHelper.parameterExists(LoanProductConstants.DISBURSED_AMOUNT_PERCENTAGE_DOWN_PAYMENT, element)) {
+                BigDecimal disbursedAmountPercentageDownPayment = this.fromApiJsonHelper
+                        .extractBigDecimalWithLocaleNamed(LoanProductConstants.DISBURSED_AMOUNT_PERCENTAGE_DOWN_PAYMENT, element);
+                baseDataValidator.reset().parameter(LoanProductConstants.DISBURSED_AMOUNT_PERCENTAGE_DOWN_PAYMENT)
+                        .value(disbursedAmountPercentageDownPayment).notLessThanMin(BigDecimal.ONE)
+                        .notGreaterThanMax(BigDecimal.valueOf(100)).scaleNotGreaterThan(Integer.valueOf(6));
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java]: Integer.valueOf(6) can be simplified: 6 ;)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java [hunk 5]
```diff
@@ -744,9 +745,38 @@ public void validateForCreate(final String json) {
...
+            } else {
+                baseDataValidator.reset().parameter(LoanProductConstants.DISBURSED_AMOUNT_PERCENTAGE_DOWN_PAYMENT).failWithCode(
+                        "required.for.enable.down.payment.true",
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java:768]: This error code seems wrong.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    @Test
+    public void loanProductUpdateWithEnableDownPaymentConfigurationTest() {
+        // Delinquency Bucket
+        final Integer delinquencyBucketId = DelinquencyBucketsHelper.createDelinquencyBucket(requestSpec, responseSpec);
+        final GetDelinquencyBucketsResponse delinquencyBucket = DelinquencyBucketsHelper.getDelinquencyBucket(requestSpec, responseSpec,
+                delinquencyBucketId);
+        // Loan Product without enable down payment configuration
+        final GetLoanProductsProductIdResponse getLoanProductsProductResponse = createLoanProduct(loanTransactionHelper,
+                delinquencyBucketId);
+        assertNotNull(getLoanProductsProductResponse);
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:97]: Please check whether the result has "enabledDownPayment=false"

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+
+        // Modify Loan Product to update enable down payment configuration
+        PutLoanProductsProductIdResponse loanProductModifyResponse = updateLoanProduct(loanTransactionHelper,
+                getLoanProductsProductResponse.getId());
+        assertNotNull(loanProductModifyResponse);
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:103]: Please check whether the result has "enabledDownPayment=true"

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+
+        GetLoansLoanIdResponse loanDetails = loanTransactionHelper.getLoanDetails(loanId.longValue());
+
+        // verify down-payment details for Loan
+        assertNotNull(loanDetails);
+        assertEquals(enableDownPayment, loanDetails.getEnableDownPayment());
+        assertEquals(0, loanDetails.getDisbursedAmountPercentageForDownPayment().compareTo(disbursedAmountPercentageForDownPayment));
+
+        // Modify Loan Product to update enable down payment configuration
+        PutLoanProductsProductIdResponse loanProductModifyResponse = updateLoanProduct(loanTransactionHelper,
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:253]: Down payment is already enabled...

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+                getLoanProductsProductResponse.getId());
+        assertNotNull(loanProductModifyResponse);
+
+        // verify Loan product configuration change
+        GetLoanProductsProductIdResponse getLoanProductsProductResponse_1 = loanTransactionHelper.getLoanProduct(loanProductId);
+        assertNotNull(getLoanProductsProductResponse_1);
+        assertEquals(enableDownPayment, getLoanProductsProductResponse_1.getEnableDownPayment());
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:260]: Down payment was already enabled, so it should be disabled in this test case, no?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my questions.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java (hunk 9 above)
Comment (ruchiD) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:260]: Hi Adam, In this test case i changed disbursedAmount percentage and not enable downpayment config. Earlier for LoanProduct it was 12.5 and loan account was created with 12.5 and then i changed percentage to 25 but loan account has percentage 12.5 with which it was created.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java (hunk 8 above)
Comment (ruchiD) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:253]: Hi Adam, In this test case i changed disbursedAmount percentage and not enable downpayment config. Earlier for LoanProduct it was 12.5 and loan account was created with 12.5 and then i changed percentage to 25 but loan account has percentage 12.5 with which it was created.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java (hunk 9 above)
Comment (adamsaghy) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:260]: I see. I was wondering whether we should have a test case where the down payment was enabled and % was set, but later it got disabled. Are we removing the % or leave it AS-IS?

Comment (ruchiD) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:260]: We are leaving downpayment AS-IS.

Comment (adamsaghy) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:260]: Is it not failing on the validation?

I recall if the down payment percentage is passed, but the down payment is not enabled, there is a validation to catch this situations, no?

Comment (ruchiD) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanProductWithDownPaymentConfigurationTest.java:260]: Yes. There is a validation for this situation. I am checking if enableDownPayment is false then downPayment percentage element should not be passed.
I am not setting that to null in db if that is so.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java (hunk 1 above)
Comment (ruchiD) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java]: done.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java (hunk 4 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java]: done

**Pull request merged**